src/shamir.rs

```diff
@@ -0,0 +1,419 @@
+//!
+//! This module contains the implementation of [`Shamir secret sharing`] using finite fields.
+//!
+//! [`Shamir secret sharing`](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)
+//!
+use crate::field::{Field, Fp31};
+use rand::Rng;
+use rand_core::RngCore;
+use std::iter::repeat_with;
+use std::num::NonZeroU8;
+use std::ops::Add;
+use thiserror::Error;
+
+/// Interpolating polynomial used to reconstruct secrets encoded with Shamir secret scheme.
+#[derive(Debug)]
+pub struct LagrangePolynomial<F> {
+    coefficients: Vec<F>,
+}
+
+impl<F: Field> LagrangePolynomial<F> {
+    /// Constructs new polynomial of a given degree.
+    ///
+    /// ## Errors
+    /// Returns an error if it fails to evaluate at least one of the coefficients
+    pub fn new(degree: NonZeroU8) -> Result<Self, Error> {
+        if u128::from(degree.get()) >= F::PRIME.into() {
+            // SAFETY: F::Prime <= u8::MAX here
+            #[allow(clippy::cast_possible_truncation)]
+            Err(Error::FieldSizeError {
+                field_size: F::PRIME.into() as u8,
+                polynomial_degree: degree.get(),
+            })
+        } else {
+            let mut coefficients = Vec::with_capacity(degree.get() as usize);
+            let n = degree.get();
+            for i in 1..=n {
+                let mut x = F::ONE;
+                let mut denom = F::ONE;
+
+                let x_i = F::from(u128::from(i));
```

Comment:
is it possible to define such bound without dyn? Adding `From<Into<u128>` bound causes a compile error for me

Reply:
Ok, it was just a theory.  I hadn't tried it.  And when implementing `From<Into<u128>>` was enough to count as implementing `From<u128>`, I thought it might be worth a shot.